src/test/java/com/microsoft/azure/relay/HybridConnectionListenerTest.java

```diff
@@ -10,64 +10,166 @@
 import java.nio.ByteBuffer;
 import java.time.Duration;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
 import org.eclipse.jetty.http.HttpStatus;
-import org.junit.AfterClass;
+import org.eclipse.jetty.websocket.api.UpgradeException;
+import org.junit.After;
+import org.junit.Before;
 import org.junit.BeforeClass;
 import org.junit.Test;
 
+import com.microsoft.azure.relay.HybridConnectionListener.ControlConnection;
+
 public class HybridConnectionListenerTest {
-	// Max # simultaneous client connections = # of cores - 1
-	// Because one thread need to be reserved for listener
 	private static final int MAX_CONNECTIONS_COUNT = Math.max(1, Runtime.getRuntime().availableProcessors() * 2);
-	private static HybridConnectionListener listener;
+	private static final AutoShutdownScheduledExecutor EXECUTOR = AutoShutdownScheduledExecutor.Create();
 	private static TokenProvider tokenProvider;
 	private static HybridConnectionClient client;
 	private static URI CONNECTION_URI;
+	private static HybridConnectionListener listener;
 	
 	@BeforeClass
 	public static void init() throws URISyntaxException, RelayException {
 		CONNECTION_URI = new URI(TestUtil.RELAY_NAMESPACE_URI + TestUtil.ENTITY_PATH);
 		tokenProvider = TokenProvider.createSharedAccessSignatureTokenProvider(TestUtil.KEY_NAME, TestUtil.KEY);
-		listener = new HybridConnectionListener(CONNECTION_URI, tokenProvider);
 		client = new HybridConnectionClient(CONNECTION_URI, tokenProvider);
-		listener.openAsync(Duration.ofSeconds(15)).join();
 	}
 	
-	@AfterClass
-	public static void cleanup() {
-		if (listener != null) {
-			listener.close();
-		}
+	@Before
+	public void beforeEachTest() {
+		listener = new HybridConnectionListener(CONNECTION_URI, tokenProvider);
+	}
+	
+	@After
+	public void afterEachTest() {
+		listener.close();
```

Comment:
>close [](start = 11, length = 5)

Is listener.close throws does the test case fail?

Reply:
if listener.close() throws it will fail that test

---
In reply to: [262216035](https://github.com/Azure/azure-relay-java/pull/28#discussion_r262216035) [](ancestors = 262216035)

Reply:
this @After attribute acts as a final block for each test, so it behaves like part of each test.

---
In reply to: [262223458](https://github.com/Azure/azure-relay-java/pull/28#discussion_r262223458) [](ancestors = 262223458,262216035)